Faster status cache

alessandrod opened the pull request:
This PR removes the global RwLock around the status cache, and introduces more granular RwLocks per-blockhash and per-slot. Additionally, it changes the internal hash tables from std HashMap to Dashmap, so that operations at the blockhash and slot level can be done only holding read locks. 

This is not the final design of A Performant Status Cache - which, I think, can make check and update go straight to 0 -  but it's a good incremental improvement. 

Results are pretty good: check_transactions is ~6x faster, and update_transaction_statuses is ~2.5x faster.

<img width="1963" alt="Screenshot 2024-11-26 at 10 57 34 pm" src="https://github.com/user-attachments/assets/a31c35cc-b5fb-4e34-99d7-89e0f4d7e9bf">

<img width="1936" alt="Screenshot 2024-11-26 at 10 59 04 pm" src="https://github.com/user-attachments/assets/6e31c6da-7d78-4823-a158-db3490f0802d">

alessandrod commented:
bench-tps against a single node

scheduler before and after

<img width="2050" alt="Screenshot 2024-12-10 at 12 33 42 am" src="https://github.com/user-attachments/assets/1a380cf8-bef2-47e9-868f-d2c5936f127d">

<img width="2048" alt="Screenshot 2024-12-10 at 12 30 24 am" src="https://github.com/user-attachments/assets/687038d7-b0fd-4397-b7e4-f6534a9ac60d">

workers

<img width="2049" alt="Screenshot 2024-12-10 at 12 30 44 am" src="https://github.com/user-attachments/assets/c6c8ebc7-7e48-4cf9-b8c1-46fd56775209">

<img width="2055" alt="Screenshot 2024-12-10 at 12 31 54 am" src="https://github.com/user-attachments/assets/226dc236-e3b5-4079-a97a-6574c6eac76c">

<img width="2055" alt="Screenshot 2024-12-10 at 12 32 04 am" src="https://github.com/user-attachments/assets/48733dfb-ef4b-4ec8-9339-2b3dd826ea24">

<img width="2050" alt="Screenshot 2024-12-10 at 12 32 14 am" src="https://github.com/user-attachments/assets/4631b150-7cfa-438e-a955-8bd0c30b3b7b">

apfitzge commented on runtime/src/status_cache.rs:
```diff
@@ -1,119 +1,159 @@
-// Store forks in a single chunk of memory to avoid another lookup.
-pub type ForkStatus<T> = Vec<(Slot, T)>;
+// The number of shards to use for ::slot_deltas. We're going to store at most MAX_CACHE_ENTRIES
+// slots, MAX * 4 gives us a low load factor guaranteeing that collisions are vey rare.
```
```suggestion
// slots, MAX * 4 gives us a low load factor guaranteeing that collisions are very rare.
```

apfitzge commented on runtime/src/status_cache.rs:
```diff
@@ -122,186 +162,361 @@ impl<T: Serialize + Clone> StatusCache<T> {
+        }
+    }
+
+    fn push(&self, item: T) {
+        self.vec.push(item);
+    }
+
+    fn iter(&self) -> impl Iterator<Item = (usize, &T)> {
+        self.vec.iter()
+    }
+}
+
+impl<T> IntoIterator for ConcurrentVec<T> {
+    type Item = T;
+    type IntoIter = boxcar::IntoIter<T>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.vec.into_iter()
+    }
+}
+
+impl<'a, T> IntoIterator for &'a ConcurrentVec<T> {
+    type Item = (usize, &'a T);
+    type IntoIter = boxcar::Iter<'a, T>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.vec.iter()
+    }
+}
```
maybe excepting serialize and deserialize, can't we just implement `Deref` to get access to the inner for all of this other stuff?

apfitzge commented on runtime/src/status_cache.rs:
```diff
@@ -1,119 +1,159 @@
+                panic!("Blockhash must exist if it exists in self.slot_deltas, slot: {slot}")
+            };
+
+            // cache_txs is self.blockhash_cache[blockhash]
```
`self.blockhash_cache` would probably be a good rename 😄

apfitzge commented on runtime/src/status_cache.rs:
```diff
@@ -1,119 +1,159 @@
+                };
+                // remove the slot from the list of forks this tx was inserted in
+                let forks = cache_tx_entry.get_mut();
+                forks.retain(|(fork_slot, _key)| *fork_slot != slot);
+                if forks.is_empty() {
```
Might be over/prematurely optimizing... but this strikes me as a bit of an odd way to do this.

AFAIK, there is no chance of duplicate fork slots in the status cache. Seems we could check the len == 1, remove entry in that case. If != 1, then we can do retain.

Actually, retaining seems a bit odd to me because let's say we have 4 forks and we remove entry at 0. we shuffle everything over by 1. They're not **huge**, so copying over may be relatively cheap, but probably could just do something like this:

```rust
                // if all the slots have been cleared or purged, we don't need to track this tx
                // anymore
                if forks.len() == 1 {
                    cache_tx_entry.remove();
                } else {
                    forks
                        .iter()
                        .position(|(fork_slot, _)| *fork_slot == slot)
                        .map(|index| forks.swap_remove(index));
                }
```

apfitzge commented on runtime/src/status_cache.rs:
```diff
@@ -122,186 +162,361 @@ impl<T: Serialize + Clone> StatusCache<T> {
         // Get the cache entry for this blockhash.
-        let (max_slot, key_index, hash_map) =
-            self.cache.entry(*transaction_blockhash).or_insert_with(|| {
+        let key_index = {
```
If we returned from this block `(key_index, key_slice)` would that perform a copy, or would compiler optimize it to effectively what you have now?

Thinking about how we can limit the scope of `MaybeUninit` so easier to maintain. Not opposed to what you have if there's **any** performance cost to my suggestion.

Basically change the inner spot where you're currently doing ptr copy to something like this:
```rust
            // Grab the key slice.
            let key_index = (*key_index).min(max_key_index);
            let key_slice = {
                let mut key_slice = MaybeUninit::<[u8; CACHED_KEY_SIZE]>::uninit();
                unsafe {
                    ptr::copy_nonoverlapping(
                        key.as_ref()[key_index..key_index + CACHED_KEY_SIZE].as_ptr(),
                        key_slice.as_mut_ptr() as *mut u8,
                        CACHED_KEY_SIZE,
                    );
                    key_slice.assume_init()
                }
            };

            // Insert the slot and tx result into the cache entry associated with
            // this blockhash and keyslice.
            let mut forks = txs.entry(key_slice).or_default();
            forks.push((slot, res.clone()));

            (key_index, key_slice)
```

now all the unsafe is in one place.
IF this introduces extra copies, let's just keep it as is. I think the compiler should be smart enough to do what we want, but probably putting too much faith in it 😢

jstarry replied:
I agree with @apfitzge that it would be better to contain `MaybeUninit` so that we aren't copying it for the two separate `assumit_init` calls because this is pretty weird and would be UB if done on a type that doesn't `impl Copy` (`[u8; 20]` _does_ implement `Copy` tho so it's fine).

Also, this optimization can be done separately from the dashmap optimizations. I'd prefer to it to be in its own PR (either before or after this one) especially since it has unsafe code

bw-solana reviewed (commented):
LGTM. One minor suggestion

apfitzge commented on runtime/src/status_cache.rs:
```diff
@@ -122,186 +162,361 @@ impl<T: Serialize + Clone> StatusCache<T> {
-    /// Get the statuses for all the root slots
+    /// Get the statuses for all the root slots.
+    ///
+    /// This is never called concurrently with add_root(), and for a slot to be a root there must be
```
No concurrent calls with this right now. Not sure we can really prevent them from being added though.

apfitzge commented on runtime/src/status_cache.rs:
```diff
@@ -122,186 +162,361 @@ impl<T: Serialize + Clone> StatusCache<T> {
-                self.cache.retain(|_, (fork, _, _)| *fork > min);
+                self.cache.retain(|_key, value| {
+                    let (max_slot, _, _) = &**value;
+                    max_slot.load(Ordering::Relaxed) > min
```
noting for myself.

`retain` will grab the **write** lock on the shard for checking the closure. This guarantees that we have unique access to this `self.cache` entry.
Concurrent access from `insert` will have completed, or will be held up wiating for this to complete on the shard; if removed, then insert would be possibly re-adding entry (that's fine).

I'm like 99% sure that Relaxed ordering here (and in `insert`) is fine because we have guaranteed the unique access via the write-lock on shard.

apfitzge reviewed (commented):
Few more comments on this round, but overall looks correct to me.

alessandrod commented:
results on a thread ripper 7965wx 64 cores / 128 threads - 10x improvement, pretty crazy

<img width="1973" alt="Screenshot 2024-12-22 at 8 38 25 pm" src="https://github.com/user-attachments/assets/b723e98c-1e8c-4b17-bf5e-fc761be1108e" />

steveluscher commented:
I hate to do anything to contribute to delaying this PR, but how about an idea that would delay it but _also_ simplify it.

You've refactored the data structure that stores this information using a concurrent `DashMap` that's capable of storing both:

* the presence of message hashes, per fork
* a transaction's result, per fork

The former is _really_ important (prevent double spends!) but requires only a [presence test](https://github.com/anza-xyz/agave/blob/9861e66f21ca58eb157df679e8ab42ced1309b51/runtime/src/bank/check_transactions.rs#L221). The latter is less risky, requires actual data storage for result `T`, and as far as I can tell is only in use by RPC nodes and `BanksServer` (ie. development & tests).

How about we _first_ split these into two cache implementations, [actually disable this write](https://github.com/anza-xyz/agave/blob/9861e66f21ca58eb157df679e8ab42ced1309b51/runtime/src/bank.rs#L2975-L2983) unless the validator is in RPC mode, and then eliminate `T` from `SmallVec<[(Slot, T); 2]>`?

alessandrod commented:
> I hate to do anything to contribute to delaying this PR, but how about an idea that would delay it but _also_ simplify it.

> How about we _first_ split these into two cache implementations, 

Sounds like something that _you_ could do after merging the current PR! 😋

I intentionally decided against changing any of the logic, to make reviews and auditing easier, and so that if we _must_ backport to 2.1 at some point (larger blocks), the diff isn't too large.

I do agree with you that the change you're proposing must be done though.

joncinque commented:
This PR contains changes to the solana sdk, which will be moved to a new repo within the next week, when v2.2 is branched from master.

Please merge or close this PR as soon as possible, or re-create the sdk changes when the new repository is ready at https://github.com/anza-xyz/solana-sdk

jstarry commented:
Are you sure it's not better to split into two caches first before merging this? Splitting into two caches is much simpler than the concurrency changes in this PR.

alessandrod commented:
> Are you sure it's not better to split into two caches first before merging this? Splitting into two caches is much simpler than the concurrency changes in this PR.

But the concurrency changes need to be made anyway right? The bad contention is in executing txs in banking/replay.

alessandrod commented:
> > But the concurrency changes need to be made anyway right? The bad contention is in executing txs in banking/replay.
> 
> Yeah, that's fair. I was just thinking that those concurrency changes are really only needed for the "recently landed tx set" cache and not for the separate "rpc tx status" cache. Shouldn't hurt to apply to both though.
> 
> Another thing I was thinking was that we could have some low risk perf improvements by no longer storing tx signatures (speeds up update) and by switching the hash algorithm to ahash (speeds up both update and get status)

we discussed this on slack but for those following here: ahash won't hurt (we could actually switch to the ad hoc hasher we recently introduced for Pubkey), but what makes the cache slow is the contention, as shown in the first screenshot in https://github.com/anza-xyz/agave/pull/3796#issuecomment-2528016752 (RwLock::read taking nearly 100% of cache time).

alessandrod commented:
> If ever this can help: have a look at this report: https://discord.com/channels/428295358100013066/560174212967432193/1364647426106851360 that, I think, involves contention on the status cache.

This is known, see https://github.com/anza-xyz/agave/pull/3796#issuecomment-2558394345

alessandrod commented:
Ok I rebased the branch! The tests are still going to fail since I had to back out the abi-example changes as those are now in a separate repo. Fixing that next.

jstarry commented on runtime/src/status_cache.rs:
```diff
@@ -120,186 +160,361 @@ impl<T: Serialize + Clone> StatusCache<T> {
+
+// Wrapper around dashmap that stores (K, Arc<V>) to minimize shard contention.
+#[derive(Debug)]
+pub struct ReadOptimizedDashMap<K, V, S>
```
By not using the read guards in `get` and `get_or_insert_with` it's possible to remove entries while they are being read, potentially introducing data race bugs.

For example, `clear_slot_entries` is problematic because it no longer runs behind a write lock. As documented, it is mostly safe because it's guaranteed that there are no threads inserting new entries for the slot being removed. But it is still possible that banking stage is writing to the status cache for a different slot while `clear_slot_entries` is running. If banking stage processed a transaction with a blockhash that was only used in a duplicate block on another fork, it's possible that `clear_slot_entries` would delete the cache entry for that blockhash while banking stage was modifying a read-only copy of that value, leading to the banking stage modifications not being applied to the status cache, and therefore leading to the leader including a transaction more than once in their blocks.

I don't think this bug would be super common but from reading `dump_then_repair_correct_slots` it does seem possible. As documented in `dump_then_repair_correct_slots`, a leader would likely *not* be building  a block descended from a duplicate block if it had recently detected one. If this leader was next in the leader schedule after the duplicate block producer, it's very possible that no other blocks included a transaction with a blockhash from a recent block yet.

Regardless of this potential bug, I don't feel comfortable adding this data structure as-is because I think that the status cache is used in complex ways that make data races difficult to reason about and there are little to no guarantees that we would notice if any assumptions about async safety are broken in the future.

alessandrod replied:
EDIT2: @apfitzge as resident banking guy, check this logic pls

jstarry commented:
> we discussed this on slack but for those following here: ahash won't hurt (we could actually switch to the ad hoc hasher we recently introduced for Pubkey), but what makes the cache slow is the contention, as shown in the first screenshot in https://github.com/anza-xyz/agave/pull/3796#issuecomment-2528016752 (RwLock::read taking nearly 100% of cache time).

I think it's clear that "`RwLock::read` taking nearly 100% of cache time" is because it's blocked on a thread which has acquired a write lock for too long due to slow modifications. Contention goes down if write lock acquisition is shorter.

alessandrod committed:
Reduce StatusCache contention.

Remove the global RwLock around the status cache, and introduce more
granular RwLocks per-blockhash and per-slot. Additionally, change the
internal hash tables from std HashMap to Dashmap, so that operations at
the blockhash and slot level can be done only holding read locks.

alessandrod commented:
Some notes I took while re-re-re-viewing the code

clean_slot_entries(slot):
- done in the replay loop from dump_then_repair_correct_slots -> purge_unconfirmed_duplicate_slot
- dump_then_repair_correct_slots is called in the ReplayStage main loop 
- ReplayStage is single threaded, dump_then_repair_correct_slots is called
  after replay_active_banks; replay_active_banks waits for the replay
  scheduler to stop so by the time it returns there are no unified scheduler
  threads replaying entries 
- never called on a slot for which the current validator is leader according
  to leader schedule => can never race with banking 
- never cleans a rooted slot (by definition if it's rooted won't be dumped)
- it can work concurrently with RPC... but I believe that interactions with
  RPC are already broken when we dump slots as per ReplayStage comment:
	
	> // TODO: What about RPC queries that had already cloned the Bank for this slot
	> // and are looking up the signature for this slot?"
	> root_bank.clear_slot_signatures(slot);

So effectively in clean_slot_entries(), the only thing to be mindful of is
that banking might be working with some of the blockhashes in
status_cache.cache that might be referenced by the txs we're removing from
the dumped slot. We use self.cache.remove_if_not_accessed_and(...is_empty()) to remove
a blockhash, which means that if banking has added (not empty) or is adding
(is accessed) txs we won't remove the blockhash.

root_slot_deltas():
- called from replay handle_votable_bank -> check_and_handle_new_root -> ... -> set_bank_forks_root 
- this too can't race with any other entry replay threads for the same reason as clean_slot_entries
- it has exclusive access to status_cache.slot_delta[the roots it accesses] since roots can't be dumped 
- it does an Arc::clone of the _contents_ (dashmap) of the roots from
  status_cache.slot_delta[the roots] and gives it to the snapshot thread 
- while the snapshot thread does its thing, purge_root() can remove old roots from
  status_cache.slot_delta, but that's ok, it only removes the values it doesn't change the content of
  the values at all (which is what the snapshot thread works with)

I don't think that there are any race concerns at all here.

add_root()/purge_roots()
- called by replay like 3 lines after root_slot_deltas()
- needs to be careful not to remove blockhashes touched by replay/banking.

Since MAX_PROCESSING_AGE = MAX_RECENT_BLOCKHASHES / 2, by the time we
purge a root all the blockhashes referenced have expired, except nonces. We
must make sure we don't remove nonces that have been used past the root
we're purging, and we do so with 
						key_in_use || max_slot.load(Ordering::Relaxed) > min
where min is the root slot we're purging and max_slot is the last slot a
blockhash was referenced in.

Additionally nonces have the nonce account data check in place that avoids replay, so I think
here we're all good.

insert(blockhash, tx, slot)
- called by the entries replay threads and by banking 
- can't run concurrently with clear_slot_entries(), see above 
- can't touch the same status_cache.slot_delta[slot] as purge_roots()
  because roots can't be inserted into
- does status_cache.slot_delta.get_or_insert(slot).get_or_insert(blockhash).push(tx) (boxcar vec)
- does status_cache.cache.get_or_insert(blockhash).txs.entry(tx).forks.push(slot) (regular vec)

So effectively, what can happen concurrently:
- purge_roots() removing blockhashes that replay/banking are inserting under
  (these can only be nonces, see above)
- replay/banking inserting concurrently under the same blockhashes (very common)
- check_transactions() called by scheduler/banking workers/replay threads
  while insert() is in progress (also common)

Because of the serialization forced by account locks, check_transactions(tx) and insert(tx) can
never be called concurrently when it matters (load_and_execute_transactions() in replay or
process_and_record_transactions() in the banking workers).

jstarry commented:
> So effectively in clean_slot_entries(), the only thing to be mindful of is
that banking might be working with some of the blockhashes in
status_cache.cache that might be referenced by the txs we're removing from
the dumped slot. We use self.cache.remove_if_not_accessed_and(...is_empty()) to remove
a blockhash, which means that if banking has added (not empty) or is adding
(is accessed) txs we won't remove the blockhash.

Yeah, I think you described this race condition correctly and the new method `remove_if_not_accessed_and` looks like a solid fix without needing a read guard. Awesome!
